### ISSUE TITLE
don't hardcode credentials for pulling from upstream registry

### ISSUE THREAD

**Issue by vikstrous** [docker/distribution#1899]
When using the proxy / pull through cache feature, the registry should use the user's permissions for pulling from the upstream registry instead of using its own hardcoded username and password. When using JWT auth we can forward the user's JWT token to the upstream registry. We might run into issues with issuer/audience validation. We need to think through this. I'm open to suggestions for how else we can achieve this.

**Comment by dmcgowan** [comment]
Agreed we need to think through this. Fundamental to this question is whether or not a mirror or proxy-cache should be trusted. We currently don't distinguish between a "public mirror" and "private cache", which makes a big difference on how we handle credentials. Even though we have designed the short-lived token system to be able to handle untrusted mirrors, as soon as the mirror is given the ability to send a `401` which the client will respond to, it can determine where credentials are send.

Perhaps the correct solution in the near term is to treat the proxy cache as trusted, in which case this proposal makes sense. It appears the docker client may not be doing the right thing now and actually sending the credentials to the cache anyway. I don't believe that is the right behavior but need to verify. In any case, this line of trust has not been clearly drawn and it may be unclear (or maybe people just don't care) about what configuring a mirror on the daemon means.

**Comment by vikstrous** [comment]
I would love to have the option to not trust the mirror, but that would probably require more configuration client side than passing everything through the mirror. Maybe we should track this feature in a separate issue as it doesn't really fit well into our current model.

Since we are trusting the mirror anyway, I think it's actually okay to trust it to proxy requests to the auth server too. That would simplify client configuration because you won't need to trust the upstream server's CA. We still need a more concrete design for this feature though, and I could use some help with that.

**Comment by boombatower** [comment]
It's been a while, but any interest in this these days? Seems like the same usecase for proxying public images applies the same to private. If the authentication is always checked via a HEAD request to the proxied registry would this not work?

**Comment by boombatower** [comment]
In fact this would also be a much simpler, intuitive solution for proxying ecr repos.
